[WIP] Gate DistributedOptimizers on RPC availability

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62774

Gates DistributedOptimizer which relies on RRef based on if RPC is available. This should enable ZeRo to work with Windows as Windows should not try to import the DIstributedOptimizer. If this works as expected we can enable the windows tests for functional/local sgd optimizers as well.

Also move `functional_optim_map` to a common place, as zero was previously reading it from DistributedOptimizer class, but we don't want zero to rely on DistributedOptimizer.

Differential Revision: [D30117838](https://our.internmc.facebook.com/intern/diff/D30117838/)

rohan-varma committed:
[WIP] Gate DistributedOptimizers on RPC availability

Gates DistributedOptimizer which relies on RRef based on if RPC is available. This should enable ZeRo to work with Windows as Windows should not try to import the DIstributedOptimizer. If this works as expected we can enable the windows tests for functional/local sgd optimizers as well.

Differential Revision: [D30117838](https://our.internmc.facebook.com/intern/diff/D30117838/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Gate DistributedOptimizers on RPC availability"


Gates DistributedOptimizer which relies on RRef based on if RPC is available. This should enable ZeRo to work with Windows as Windows should not try to import the DIstributedOptimizer. If this works as expected we can enable the windows tests for functional/local sgd optimizers as well.

Differential Revision: [D30117838](https://our.internmc.facebook.com/intern/diff/D30117838/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Gate DistributedOptimizers on RPC availability"


Gates DistributedOptimizer which relies on RRef based on if RPC is available. This should enable ZeRo to work with Windows as Windows should not try to import the DIstributedOptimizer. If this works as expected we can enable the windows tests for functional/local sgd optimizers as well.

Also move `functional_optim_map` to a common place, as zero was previously reading it from DistributedOptimizer class, but we don't want zero to rely on DistributedOptimizer.

Differential Revision: [D30117838](https://our.internmc.facebook.com/intern/diff/D30117838/)

[ghstack-poisoned]

rohan-varma commented:
@mrshenli this implements your suggestion in https://github.com/pytorch/pytorch/issues/62137.

mrshenli reviewed (approved):
Thanks @rohan-varma! This looks great!

rohan-varma committed:
Update on "[WIP] Gate DistributedOptimizers on RPC availability"


Gates DistributedOptimizer which relies on RRef based on if RPC is available. This should enable ZeRo to work with Windows as Windows should not try to import the DIstributedOptimizer. If this works as expected we can enable the windows tests for functional/local sgd optimizers as well.

Also move `functional_optim_map` to a common place, as zero was previously reading it from DistributedOptimizer class, but we don't want zero to rely on DistributedOptimizer.

Differential Revision: [D30117838](https://our.internmc.facebook.com/intern/diff/D30117838/)

[ghstack-poisoned]
